library/crypto/sha512.c

```diff
@@ -260,22 +260,28 @@ sha512_file (int fd, off_t offset, size_t count, void *resblock)
           if (sum == BLOCKSIZE)
             break;
 
-          if (n == 0) // eof
-              goto process_partial_block;
+          if (n == 0) { /* eof */
+              if (total + sum < count) {
+                  free (buffer);
+                  return -ESPIPE;
+              } else {
+                  goto process_partial_block;
+              }
+          }
         }
 
-      /* Process buffer with BLOCKSIZE bytes.  Note that
-                        BLOCKSIZE % 128 == 0
-       */
-      if (total + BLOCKSIZE < count) {
-          total += BLOCKSIZE;
+      if (total + sum < count) {
```

Comment:
The fact, that total number of bytes read from fd is very small (thus `count - total > BLOCKSIZE`) is not a reason for bug, this is a consequence. What was the reason?
How did it happen, that we have a record, which size is large enough to span outside of the datafile?
Was it data corruption? Do we have any logs for this record?

This change is valid, but we do have to understand which steps lead to this situation and subsequent crash.

*I wrote this function and sha512_file_ctx() during mmap -> pread transition in eblob. And I missed this bug, because did not wrote tests.*

Yes, this requires tests both in elliptics and eblob.